src/ensembl/doc/architecture/decisions/0003-state-management.md

```diff
@@ -14,8 +14,14 @@ The state of data and UI will grow in the application as the development goes on
 
 [Redux](https://redux.js.org/) will be used for the state management. It is a centralised state management container, that can handle the state for data and UI of the application.
 
+Using React.js state only, instead of any state management library, was considered. However, this would make it difficult to share state between sibling components and manage global state as well.
+
+Also, Mobx was discussed about. However, the magic nature of how things are done with it compared to Redux made the team decide not to go ahead with it.
```

Comment:
My attempt:

There are various ways of state management in a React application. React itself provides two mechanisms:

- component's local state (useful for tracking transient localised state that most parts of the application are not concerned with)
- React context (useful for sharing state between unrelated components)

Both of these mechanisms, while useful, are generally insufficient for tracking state in a large application, which resulted in proliferation of numerous third-party libraries for state management. Of those, two most notable are:

- redux
- mobX

These two libraries represent two different philosophies of state management (and programming in general). Redux follows the functional programming approach, in which state itself is immutable, and state changes are a series of new copies of state produced in response to message events ("actions") dispatched within the store. In MobX, state is a mutable object that is changed via getters and setters (in latest implementations, proxies), which also notify subscriptions of the change.

While MobX's approach is generally regarded as easier to use, we feel that it also makes this library more "magical" and therefore more difficult to debug as compared to Redux. Also, Redux's reliance on events makes it a convenient target for plugging various kinds of middleware (i.e. analytics middleware, redux-saga or redux-observable) for better encapsulation of state-dependent logic.

Reply:
Thanks. This looks great. I'll add it to the ADR.